compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+            EUpgradeTemplateChangedAgreement $ NM.name $ present template
+    -- TODO: Check that return type of a choice is compatible
+    pure ()
+    where
+        extractFuncFromFuncThis expr
+            | ETmApp{..} <- expr
+            , EVal qualEvn <- tmappFun
+            , EVar (ExprVarName "this") <- tmappArg
+            = Just (qualObject qualEvn)
+            | otherwise
+            = Nothing
+        extractFuncFromFuncThisArg expr
+            | outer@ETmApp{} <- expr
+            , EVar (ExprVarName "arg") <- tmappArg outer
+            , inner@ETmApp{} <- tmappFun outer
+            , EVar (ExprVarName "this") <- tmappArg inner
+            , EVal qualEvn <- tmappFun inner
+            = Just (qualObject qualEvn)
+            | otherwise
+            = Nothing
+        extractFuncFromCaseFuncThis expr
+            | ECase{..} <- expr
+            = extractFuncFromFuncThis casScrutinee
+            | otherwise
+            = Nothing
+        resolveExpression field expr module_ =
+            case expr of
+              Nothing -> error ("checkTemplate: Could not extract a proper " ++ field ++ ", the structure of the expression must be wrong.")
+              Just evn ->
+                case NM.lookup evn (moduleValues module_) of
+                    Nothing -> error ("checkTemplate: Trying to get definition of " ++ T.unpack (unExprValName evn) ++ " but it is not defined!")
+                    Just defValue -> dvalBody defValue
+        throwIfDifferent field exprs err = do
+            let resolvedExprs = resolveExpression field <$> exprs <*> module_
+            let exprsMatch = foldU alphaExpr $ fmap removeLocations resolvedExprs
+            unless exprsMatch (throwWithContext err)
+        mapENilToNothing (Just (LF.ENil (LF.TBuiltin LF.BTParty))) = Nothing
+        mapENilToNothing e = e
+
+checkDefDataType :: UpgradedRecordOrigin -> Upgrading LF.DefDataType -> Maybe Error
+checkDefDataType origin datatype = do
+    case fmap dataCons datatype of
+      Upgrading { past = DataRecord past, present = DataRecord present } -> checkFields origin (Upgrading {..})
+      Upgrading { past = DataVariant {}, present = DataVariant {} } -> Nothing
+      Upgrading { past = DataEnum {}, present = DataEnum {} } -> Nothing
+      Upgrading { past = DataInterface {}, present = DataInterface {} } -> Nothing
+      _ -> Just (EUpgradeMismatchDataConsVariety (dataTypeCon (past datatype)))
+
+checkFields :: UpgradedRecordOrigin -> Upgrading [(FieldName, Type)] -> Maybe Error
+checkFields origin fields =
+    let (deleted, existing, new) = extractDelExistNew $ HMS.fromList <$> fields
+    in
+    if not (HMS.null deleted) then
+        Just (EUpgradeRecordFieldsMissing origin)
+    -- If a field from the upgraded package has had its type changed
+    else if any matchingFieldDifferentType existing then
+        Just (EUpgradeRecordFieldsExistingChanged origin)
```

Comment:
> To be clear, you did mention in another thread that you'll implement the full check for type compatibility in a later PR, so this is just to point out another place where it will be needed

Cheers - issue for the PR to come: https://github.com/digital-asset/daml/issues/17611